File: Lab 8 Network Processor Integration/Emiliano/sw/send_test.py
import socket
import struct
import sys

# Configuration
# Node 0 IP address as requested
DEST_IP = "10.0.0.2" # The IP address of NetFPGA 
PORT = 5000
TARGET_VALUE = 5      # The initial value to be incremented by the CPU

# Your net_CPU_test.c uses PAYLOAD_OFFSET 8.
# In your 64-bit architecture (8 bytes per word):
# Word 0 starts at Byte 0
# Word 8 starts at Byte 64 (8 * 8)
OFFSET = 64

def send_packet():
    try:
        # Create a UDP socket
        sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
        
        # Build the payload: 
        # 1. 64 bytes of 'A' as padding to reach word index 8
        # 2. The 64-bit integer TARGET_VALUE packed in Little-Endian format (<Q)
        padding = b'A' * OFFSET
        data = struct.pack('<Q', TARGET_VALUE)
        payload = padding + data
        
        print(f"Sending UDP packet to {DEST_IP}:{PORT}...")
        print(f"Original value at byte offset {OFFSET}: {TARGET_VALUE}")
        
        sock.sendto(payload, (DEST_IP, PORT))
        print("Packet sent successfully.")
        
        print("\nHow to verify:")
        print(f"1. On NetFPGA (10.0.0.2), run:")
        print(f"   sudo tcpdump -i eth1 -XX udp port {PORT}")
        print(f"2. Look for the value at the end of the packet.")
        print(f"   It should have changed from 0x05 to 0x69 (105 decimal).")
        
    except Exception as e:
        print(f"Error: {e}")
    finally:
        sock.close()

if __name__ == "__main__":
    send_packet()

File: DPU/sw/NetFPGA Scripts/send_test.pl
#!/usr/bin/perl -w

use strict;
use IO::Socket::INET;

# --- Configuration ---
my $DEST_IP = "10.0.0.2";
my $PORT    = 5000;
my $VALUE   = 5;      # Initial value to be incremented
# OFFSET 78 + ~50 bytes of headers = 128 bytes.
# 128 bytes / 8 bytes per word = Word Index 16.
# CPU accesses Word Index 16 using address offset 64 (16 * 4).
my $OFFSET  = 78;     

print "--- EE 533 Lab 8: UDP Packet Sender (Perl) ---\n";
print "Target IP: $DEST_IP\n";
print "Target Port: $PORT\n";
print "Value at offset $OFFSET: $VALUE\n";

# 1. Create the payload
# 'A' x 64 creates 64 bytes of padding
# We pack two 32-bit little-endian longs (L) because old Perl doesn't support 'Q'
# This creates a 64-bit little-endian unsigned integer.
my $payload = ('A' x $OFFSET) . pack('L L', $VALUE, 0);

# 2. Create the UDP socket
my $sock = IO::Socket::INET->new(
    PeerAddr => $DEST_IP,
    PeerPort => $PORT,
    Proto    => 'udp'
) or die "Could not create socket: $!\n";

# 3. Send the packet
$sock->send($payload) or die "Send failed: $!\n";

print "Packet sent successfully.\n";

$sock->close();
